File: symphony/app/fbcnms-projects/inventory/app/common/Project.js
/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

import type {Location} from './Location';
import type {NamedNode} from './EntUtils';
import type {ProjectTemplateNodesQuery} from './__generated__/ProjectTemplateNodesQuery.graphql';
import type {Property} from './Property';
import type {PropertyType} from './PropertyType';
import type {WorkOrder} from './WorkOrder';

import {graphql} from 'relay-runtime';
import {useLazyLoadQuery} from 'react-relay/hooks';

export type ProjectType = {
  id: string,
  name: string,
  propertyTypes: Array<PropertyType>,
};

export type Project = {
  id: string,
  type: ?ProjectType,
  name: string,
  description: ?string,
  location: ?Location,
  creatorId: ?string,
  properties: Array<Property>,
  workOrders: Array<WorkOrder>,
  numberOfWorkOrders: number,
};

const projectTemplateNodesQuery = graphql`
  query ProjectTemplateNodesQuery {
    projectTypes {
      edges {
        node {
          id
          name
        }
      }
    }
  }
`;

export type ProjectTemplateNode = $Exact<NamedNode>;

export function useProjectTemplateNodes(): $ReadOnlyArray<ProjectTemplateNode> {
  const response = useLazyLoadQuery<ProjectTemplateNodesQuery>(
    projectTemplateNodesQuery,
  );
  const projectTemplatesData = response.projectTypes?.edges || [];
  const projectTemplates = projectTemplatesData
    .map(p => p.node)
    .filter(Boolean);
  return projectTemplates;
}

File: symphony/graph/ent/schema/project.go
// Copyright (c) 2004-present Facebook All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package schema

import (
	"github.com/facebookincubator/ent"
	"github.com/facebookincubator/ent/schema/edge"
	"github.com/facebookincubator/ent/schema/field"
	"github.com/facebookincubator/ent/schema/index"
	"github.com/facebookincubator/symphony/graph/authz"
)

// ProjectType defines the project type schema.
type ProjectType struct {
	schema
}

// Fields returns project fields.
func (ProjectType) Fields() []ent.Field {
	return []ent.Field{
		field.String("name").
			NotEmpty().
			Unique(),
		field.Text("description").
			Optional().
			Nillable(),
	}
}

// Edges return project type edges.
func (ProjectType) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("projects", Project.Type),
		edge.To("properties", PropertyType.Type),
		edge.To("work_orders", WorkOrderDefinition.Type),
	}
}

// Policy returns project type policy.
func (ProjectType) Policy() ent.Policy {
	return authz.NewPolicy(
		authz.WithMutationRules(
			authz.ProjectTypeWritePolicyRule(),
		),
	)
}

// Project defines the project schema.
type Project struct {
	schema
}

// Fields returns project fields.
func (Project) Fields() []ent.Field {
	return []ent.Field{
		field.String("name").
			NotEmpty().
			Unique(),
		field.Text("description").
			Optional().
			Nillable(),
	}
}

// Edges returns project edges.
func (Project) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("type", ProjectType.Type).
			Ref("projects").
			Unique().
			Required(),
		edge.To("location", Location.Type).
			Unique(),
		edge.To("comments", Comment.Type),
		edge.To("work_orders", WorkOrder.Type),
		edge.To("properties", Property.Type),
		edge.To("creator", User.Type).
			Unique(),
	}
}

// Indexes return project indexes.
func (Project) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("name").
			Edges("type").
			Unique(),
	}
}

// EquipmentPortDefinition defines the equipment port definition schema.
type WorkOrderDefinition struct {
	schema
}

// Fields returns equipment port definition fields.
func (WorkOrderDefinition) Fields() []ent.Field {
	return []ent.Field{
		field.Int("index").
			Optional(),
	}
}

// Edges returns equipment port definition edges.
func (WorkOrderDefinition) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("type", WorkOrderType.Type).
			Unique(),
		edge.From("project_type", ProjectType.Type).
			Ref("work_orders").
			Unique(),
	}
}
